kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/ImmutableCollection.kt

```diff
@@ -16,24 +16,76 @@
 
 package kotlinx.collections.immutable
 
+/**
+ * A generic immutable collection of elements. Methods in this interface support only read-only access to the collection;
+ * modification operations are supported through the [PersistentCollection] interface.
+ *
+ * Implementors of this interface take responsibility to be immutable,
+ * that is objects reachable from an instance of the implementor should never be mutated.
+ *
+ * @param E the type of elements contained in the collection. The immutable collection is covariant on its element type.
+ */
 public interface ImmutableCollection<out E>: Collection<E>
 
+/**
+ * A generic persistent collection of elements that supports adding and removing elements.
+ * Modification operations return new instances of the persistent collection with the modification applied.
+ *
+ * @param E the type of elements contained in the collection. The persistent collection is covariant on its element type.
+ */
 public interface PersistentCollection<out E> : ImmutableCollection<E> {
+    /**
+     * Returns a new instance of the persistent collection with the specified [element] added.
+     */
     fun add(element: @UnsafeVariance E): PersistentCollection<E>
 
+    /**
+     * Returns a new instance of the persistent collection with the specified elements added.
+     */
     fun addAll(elements: Collection<@UnsafeVariance E>): PersistentCollection<E>
 
+    /**
+     * Returns a new instance of the persistent collection with a single instance of the specified [element],
```

Comment:
"with element" and "removed" quite contradict each other.
Proposing "having a single element equal to the specified [element] removed"

Reply:
Inheritors where the order is predictable could override docs with "having _the first_ element ..."

Reply:
Here "with" preposition is used as a synonym of "by"/"via" indicating method of getting the new instance.